### ISSUE TITLE
Cannot allocate list/map, as_string

### ISSUE THREAD

**Issue by solidfoxrock** [aerospike/aerospike-client-php#43]
Hi,

I had some problems saving some random big associate arrays, the errors are `Cannot allocate list/map` or `Cannot allocate as_string`. If my bin is just simply a string or list of strings then I can save very big objects (say 10MB). But if my bin is an array of arrays then I have the errors with much smaller objects, like around a few hundreds of KB.

From the error messages I found that it's related to some kinds of static pool and there is `AS_MAX_STORE_SIZE 1024` in `aerospike_common.h`, I tried to increase this value and rebuild the extension, but it didn't help. If somone can help to explain me more about this I would be very appreciated.

Below is a script to reproduce the problem:

```php
<?php

$HOST = 'docker';
$PORT = 3000;
$config = [ 'hosts' => [ [ 'addr' => $HOST, 'port' => $PORT ] ] ];

$db = new Aerospike($config, false);
if (!$db->isConnected()) {
    echo "Could not connect to host $HOST:$PORT [{$db->errorno()}]: {$db->error()}";
    exit(1);
}


// Case 1: `value` bin is an array of strings
$k = 'mykey1';
$bins = ['key' => $k, 'value' => []];
for ($i = 0; $i < 100000; $i++) { 
    $bins['value'][] = 'foo';
}
put($k, $bins, $db);

// Case 2: `value` bin is an array of simple arrays
$k = 'mykey2';
$bins = ['key' => $k, 'value' => []];
for ($i = 0; $i < 20000; $i++) { 
    $bins['value'][] = ['foo'];
}
put($k, $bins, $db);

// Case 3: `value` bin is an array of associate arrays
$k = 'mykey3';
$bins = ['key' => $k, 'value' => []];
for ($i = 0; $i < 20000; $i++) { 
    $bins['value'][] = ['foo' => 'bar'];
}
put($k, $bins, $db);



$db->close();

function put($key, $bins, $db)
{
    echo 'Saving ' . strlen(json_encode($bins)) . '-byte object ' . $key . PHP_EOL;
    $status = $db->put($db->initKey('test', 'test', $key), $bins);
    check($status, $db);
}

function check($status, $db)
{
    if ($status == Aerospike::OK) {
        echo 'Success', PHP_EOL;
    } else {
        echo "Error: [{$db->errorno()}] {$db->error()}", PHP_EOL;
    }
}
```

**Output**

```
Saving 600026-byte object mykey1
Success
Saving 160026-byte object mykey2
Error: [-1] Cannot allocate list/map
Saving 280026-byte object mykey3
Error: [-1] Cannot allocate as_string
```

Thanks,
Kien

**Comment by rbotzer** [comment]
Your records are bound by a uniform size limit that by default is 128K and can be [configured](https://discuss.aerospike.com/t/block-size-update-how-do-i-increase-block-write-size/681) up to 1MB. This has to do with efficiently writing and reading records to storage, as Aerospike does not fragment a record. The [FAQ](https://www.aerospike.com/docs/guide/FAQ.html) and [discussion forum](https://discuss.aerospike.com/t/write-block-size-and-max-record-size/1171) have more on the topic.

**Comment by solidfoxrock** [comment]
Hi @rbotzer, is `write-block-size` only applicable to HDD / SSD storage? I'm using memory storage engine.

```
namespace test {
    replication-factor 2
    memory-size 4G
    default-ttl 30d # 30 days, use 0 to never expire/evict.

    storage-engine memory
}
```

**Comment by rbotzer** [comment]
Ah, you're not writing to a device. Okay, I'll look into it.

**Comment by pesha** [comment]
We were experiencing similar problem (see [topic on forum](https://discuss.aerospike.com/t/cannot-write-into-aerospike-cannot-allocate-as-string/1316)). The problem will occur when the array contains more than 1024 objects. In that case the serialization callback runs for each object separately and obviously there is 1024 limit. So we have implemented workaround when we are serializing the whole array ahead of internal serialization.

**Comment by rbotzer** [comment]
As a first step we increased the stack allocations for list and map types to 4096 on Linux and 2560 on OS X, as fits with each OS. That should cover the case of records in namespaces that are persisted to SSDs.

Further changes would require a combination of stack allocations and heap allocations, which opens up the door for all sorts of bugs. Since this is an edge use case for in-memory namespaces and only for a very large number of list and map elements, I'm not sure it's worth the risk.

This is included with [release 3.3.16](https://github.com/aerospike/aerospike-client-php/releases/tag/3.3.16). Please verify.
